src/markov/mc_tools.jl

```diff
@@ -291,189 +291,155 @@ todense(::Type, A::Array) = A
 
-    ts_length, k = size(X)
-
     for i in 1:k
+        i_state = real_init[i]
         for t in 1:ts_length-1
-            X[t+1, i] = draw(P_dist[X[t,i]])
+            i_state = draw(P_dist[i_state])
+            X[t+1, i] = mc.state_values[i_state]
         end
     end
     X
+
 end
 
+# ------------------------ #
+# simulate_indices methods #
+# ------------------------ #
+
 """
-Simulate time series of state transitions of the Markov chain `mc`.
+Simulate one sample path of the Markov chain `mc`.
+The resulting vector has the indices of the state values of `mc` as elements.
 
-##### Arguments
+### Arguments
 
 - `mc::MarkovChain` : MarkovChain instance.
-- `ts_length::Int` : Length of each simulation.
-- `init_state::Int(rand(1:n_states(mc)))` : Initial state.
+- `ts_length::Int` : Length of simulation
+- `;init::Int=rand(1:n_states(mc))` : Initial state 
 
-##### Returns
+### Returns
 
-- `x::Vector`: A vector of transition indices for a single simulation.
+- `X::Vector{Int}` : Vector containing the sample path, with length 
+ts_length
 """
-function simulation(mc::MarkovChain, ts_length::Int,
-                    init_state::Int=rand(1:n_states(mc)))
-    simulate(mc, ts_length, init_state; num_reps=1)[:, 1]
-end
 
-# ----------------------- #
-# simulate_values methods #
-# ----------------------- #
-
-function simulate_values(mc::MarkovChain,
-                         ts_length::Int,
-                         init_state::Vector{Int};
-                         num_reps::Int=1)
-    k = length(init_state)*num_reps
-    X = Array(eltype(mc.state_values), ts_length, k)
-    init_state = repmat(init_state, num_reps)
-    X[1, :] = mc.state_values[init_state]
-
-    simulate_values!(mc, init_state, X)
-    return X
+function simulate_indices(mc::MarkovChain, ts_length::Int;
+                          init::Int=rand(1:n_states(mc)))
+    X = Array(Int, ts_length, 1)
+    simulate_indices!(mc, X; init=init)
+    return vec(X)
 end
 
-function simulate_values(mc::MarkovChain, ts_length::Int, init_state::Int;
-                         num_reps::Int=1)
-    simulate_values(mc, ts_length, [init_state], num_reps=num_reps)
-end
 
-function simulate_values(mc::MarkovChain, ts_length::Int; num_reps::Int=1)
-    init_state = rand(1:size(mc.p, 1), num_reps)
-    simulate_values(mc, ts_length, init_state)
-end
+"""
+Fill `X` with sample paths of the Markov chain `mc` as columns.
+The resulting matrix has the indices of the state values of `mc` as elements.
+
+### Arguments
+
+- `mc::MarkovChain` : MarkovChain instance.
+- `X::Matrix{Int}` : Preallocated matrix to be filled with indices
+of the sample paths of the Markov chain `mc`.
+- `;init=rand(1:n_states(mc))` : Can be one of the following
+    - blank: random initial condition for each chain
+    - scalar: same initial condition for each chain
+    - vector: cycle through the elements, applying each as an
+      initial condition
+"""
+
+function simulate_indices!(mc::MarkovChain, X::Matrix{Int};
```

Comment:
Same comment about having X come first. I also think we can loosen its type constraint so the `X` part of the function signature is

```Julia
simulate_indices!{T<:Integer}(X::Union{AbstractVector{T},AbstractMatrix{T}}, ...)
```

This will allow for all Integer subtypes and filling an vector-or-matrix-like thing.